File: leetcode/archived/C++/Swap_nodes_in_pairs.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode* swapPairs(ListNode* head) {
        if(!head)
            return NULL;
        if(!(head->next))
            return head;
        ListNode *node1 = NULL, *node2 = NULL, *temp = NULL;
        node1 = head;
        node2 = head->next;
        
        temp = node2->next;
        node2->next = node1;
        node1->next = swapPairs(temp);
        
        return node2;
    }
};

File: leetcode/linklist/Swap Nodes in Pairs.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
 func swapPairs(head *ListNode) *ListNode {
    dummyNode := &ListNode{0, head}
    pre := dummyNode
    for ;pre != nil && pre.Next != nil && pre.Next.Next != nil;{
        node1 := pre.Next
        node2 := pre.Next.Next
        tmp := node2.Next

        pre.Next = node2 
        node2.Next = node1
        node1.Next = tmp 
        
        pre = node1
    }
    return dummyNode.Next
}